# Diamonds Dataset Analysis

## Data Modification: EDA, Cleaning, and Transformation

This notebook focuses on exploring, cleaning, and transforming the diamonds dataset.


## 1. Setup

Import required libraries and load the diamonds dataset.


In [1]:
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# Set style for better-looking plots
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (10, 6)

# Load the diamonds dataset
diamonds = sns.load_dataset("diamonds")

print("Dataset loaded successfully!")
print(f"Shape: {diamonds.shape}")


Dataset loaded successfully!
Shape: (53940, 10)


## 2. Exploratory Data Analysis (EDA)

Let's explore the dataset structure, data types, and basic statistics.


In [2]:
# Display basic information about the dataset
print("=" * 60)
print("DATASET SHAPE")
print("=" * 60)
print(f"Rows: {diamonds.shape[0]:,}")
print(f"Columns: {diamonds.shape[1]}")
print()

print("=" * 60)
print("FIRST FEW ROWS")
print("=" * 60)
diamonds.head(10)


DATASET SHAPE
Rows: 53,940
Columns: 10

FIRST FEW ROWS


,carat,cut,color,clarity,depth,table,price,x,y,z
0,0.23,Ideal,E,SI2,61.5,55.0,326,3.95,3.98,2.43
1,0.21,Premium,E,SI1,59.8,61.0,326,3.89,3.84,2.31
2,0.23,Good,E,VS1,56.9,65.0,327,4.05,4.07,2.31
3,0.29,Premium,I,VS2,62.4,58.0,334,4.20,4.23,2.63
4,0.31,Good,J,SI2,63.3,58.0,335,4.34,4.35,2.75
5,0.24,Very Good,J,VVS2,62.8,57.0,336,3.94,3.96,2.48
6,0.24,Very Good,I,VVS1,62.3,57.0,336,3.95,3.98,2.47
7,0.26,Very Good,H,SI1,61.9,55.0,337,4.07,4.11,2.53
8,0.22,Fair,E,VS2,65.1,61.0,337,3.87,3.78,2.49
9,0.23,Very Good,H,VS1,59.4,61.0,338,4.00,4.05,2.39


In [3]:
# Data types and basic info
print("=" * 60)
print("DATA TYPES AND INFO")
print("=" * 60)
print(diamonds.dtypes)
print()
print(diamonds.info())


DATA TYPES AND INFO
carat       float64
cut        category
color      category
clarity    category
depth       float64
table       float64
price         int64
x           float64
y           float64
z           float64
dtype: object

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   carat    53940 non-null  float64 
 1   cut      53940 non-null  category
 2   color    53940 non-null  category
 3   clarity  53940 non-null  category
 4   depth    53940 non-null  float64 
 5   table    53940 non-null  float64 
 6   price    53940 non-null  int64   
 7   x        53940 non-null  float64 
 8   y        53940 non-null  float64 
 9   z        53940 non-null  float64 
dtypes: category(3), float64(6), int64(1)
memory usage: 3.0 MB
None


In [4]:
# Check for missing values
print("=" * 60)
print("MISSING VALUES")
print("=" * 60)
missing = diamonds.isnull().sum()
missing_pct = (missing / len(diamonds)) * 100
missing_df = pd.DataFrame({
    'Missing Count': missing,
    'Missing Percentage': missing_pct
})
missing_df = missing_df[missing_df['Missing Count'] > 0].sort_values('Missing Count', ascending=False)
if len(missing_df) > 0:
    print(missing_df)
else:
    print("No missing values found!")


MISSING VALUES
No missing values found!


In [5]:
# Unique values for categorical columns
print("=" * 60)
print("UNIQUE VALUES FOR CATEGORICAL COLUMNS")
print("=" * 60)
categorical_cols = diamonds.select_dtypes(include=['object', 'category']).columns
for col in categorical_cols:
    unique_vals = diamonds[col].unique()
    print(f"\n{col}:")
    print(f"  Unique count: {len(unique_vals)}")
    print(f"  Values: {list(unique_vals)}")
    print(f"  Value counts:")
    print(diamonds[col].value_counts().to_string())


UNIQUE VALUES FOR CATEGORICAL COLUMNS

cut:
  Unique count: 5
  Values: ['Ideal', 'Premium', 'Good', 'Very Good', 'Fair']
  Value counts:
cut
Ideal        21551
Premium      13791
Very Good    12082
Good          4906
Fair          1610

color:
  Unique count: 7
  Values: ['E', 'I', 'J', 'H', 'F', 'G', 'D']
  Value counts:
color
G    11292
E     9797
F     9542
H     8304
D     6775
I     5422
J     2808

clarity:
  Unique count: 8
  Values: ['SI2', 'SI1', 'VS1', 'VS2', 'VVS2', 'VVS1', 'I1', 'IF']
  Value counts:
clarity
SI1     13065
VS2     12258
SI2      9194
VS1      8171
VVS2     5066
VVS1     3655
IF       1790
I1        741


In [6]:
# Basic statistics for numerical columns
print("=" * 60)
print("NUMERICAL SUMMARY STATISTICS")
print("=" * 60)
diamonds.describe()


NUMERICAL SUMMARY STATISTICS


,carat,depth,table,price,x,y,z
count,53940.000000,53940.000000,53940.000000,53940.000000,53940.000000,53940.000000,53940.000000
mean,0.797940,61.749405,57.457184,3932.799722,5.731157,5.734526,3.538734
std,0.474011,1.432621,2.234491,3989.439738,1.121761,1.142135,0.705699
min,0.200000,43.000000,43.000000,326.000000,0.000000,0.000000,0.000000
25%,0.400000,61.000000,56.000000,950.000000,4.710000,4.720000,2.910000
50%,0.700000,61.800000,57.000000,2401.000000,5.700000,5.710000,3.530000
75%,1.040000,62.500000,59.000000,5324.250000,6.540000,6.540000,4.040000
max,5.010000,79.000000,95.000000,18823.000000,10.740000,58.900000,31.800000


## 3. Data Cleaning

Handle missing values, set proper data types, and rename columns if needed.


In [7]:
# Create a copy for cleaning
df_clean = diamonds.copy()

print("Starting data cleaning process...")
print(f"Original shape: {df_clean.shape}")


Starting data cleaning process...
Original shape: (53940, 10)


In [8]:
# Handle missing values (if any)
# Since diamonds dataset typically has no missing values, we'll document the approach
if df_clean.isnull().sum().sum() > 0:
    print("Missing values detected. Handling them...")
    # For numerical columns, we could use median or mean
    # For categorical columns, we could use mode or 'Unknown'
    # For this dataset, we'll document the approach but likely won't need it
    print("Missing value handling strategy:")
    print("  - Numerical: Fill with median (robust to outliers)")
    print("  - Categorical: Fill with mode or 'Unknown'")
else:
    print("No missing values to handle. Dataset is complete.")


No missing values to handle. Dataset is complete.


In [9]:
# Set proper data types
# Convert categorical columns to category type for better performance and memory usage
print("\nSetting proper data types...")

# Categorical columns that should be ordered
ordered_categories = {
    'cut': ['Fair', 'Good', 'Very Good', 'Premium', 'Ideal'],
    'color': ['J', 'I', 'H', 'G', 'F', 'E', 'D'],  # D is best, J is worst
    'clarity': ['I1', 'SI2', 'SI1', 'VS2', 'VS1', 'VVS2', 'VVS1', 'IF']  # IF is best, I1 is worst
}

for col, order in ordered_categories.items():
    if col in df_clean.columns:
        df_clean[col] = pd.Categorical(df_clean[col], categories=order, ordered=True)
        print(f"  - {col}: Converted to ordered categorical")

# Check for any other object columns that should be categorical
other_categorical = df_clean.select_dtypes(include=['object']).columns
for col in other_categorical:
    if col not in ordered_categories:
        df_clean[col] = df_clean[col].astype('category')
        print(f"  - {col}: Converted to categorical")

print("\nData types after conversion:")
print(df_clean.dtypes)


Setting proper data types...
  - cut: Converted to ordered categorical
  - color: Converted to ordered categorical
  - clarity: Converted to ordered categorical

Data types after conversion:
carat       float64
cut        category
color      category
clarity    category
depth       float64
table       float64
price         int64
x           float64
y           float64
z           float64
dtype: object


In [10]:
# Check for outliers in numerical columns
print("\nChecking for potential outliers...")
numerical_cols = df_clean.select_dtypes(include=[np.number]).columns

for col in numerical_cols:
    Q1 = df_clean[col].quantile(0.25)
    Q3 = df_clean[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = ((df_clean[col] < lower_bound) | (df_clean[col] > upper_bound)).sum()
    if outliers > 0:
        print(f"  - {col}: {outliers:,} potential outliers ({outliers/len(df_clean)*100:.2f}%)")
    else:
        print(f"  - {col}: No outliers detected")

print("\nNote: Outliers in diamond data may be legitimate (e.g., very large or expensive diamonds).")
print("We'll keep them unless they represent data entry errors.")


Checking for potential outliers...
  - carat: 1,889 potential outliers (3.50%)
  - depth: 2,545 potential outliers (4.72%)
  - table: 605 potential outliers (1.12%)
  - price: 3,540 potential outliers (6.56%)
  - x: 32 potential outliers (0.06%)
  - y: 29 potential outliers (0.05%)
  - z: 49 potential outliers (0.09%)

Note: Outliers in diamond data may be legitimate (e.g., very large or expensive diamonds).
We'll keep them unless they represent data entry errors.


## 4. Data Transformation

Create helpful derived columns (ratios, bins, flags) and summarize the data.


In [11]:
# Create helpful derived columns
print("Creating derived columns...")

# 1. Price per carat (important metric for diamond value)
df_clean['price_per_carat'] = df_clean['price'] / df_clean['carat']
print("  - Created 'price_per_carat': price divided by carat weight")

# 2. Volume (approximate, using x * y * z)
df_clean['volume'] = df_clean['x'] * df_clean['y'] * df_clean['z']
print("  - Created 'volume': x * y * z (cubic mm)")

# 3. Depth percentage (already exists, but let's verify it's correct)
# depth = (z / mean(x, y)) * 100
df_clean['depth_calculated'] = (df_clean['z'] / ((df_clean['x'] + df_clean['y']) / 2)) * 100
print("  - Created 'depth_calculated': calculated depth percentage")

# 4. Table percentage (already exists, but let's verify)
# table = (table width / average of x and y) * 100
print("  - 'table' column already exists as table percentage")

# 5. Size category based on carat
df_clean['size_category'] = pd.cut(
    df_clean['carat'],
    bins=[0, 0.5, 1.0, 2.0, float('inf')],
    labels=['Small', 'Medium', 'Large', 'Very Large']
)
print("  - Created 'size_category': Small (<0.5), Medium (0.5-1.0), Large (1.0-2.0), Very Large (>2.0)")

# 6. Price category
df_clean['price_category'] = pd.qcut(
    df_clean['price'],
    q=4,
    labels=['Budget', 'Mid-Range', 'Premium', 'Luxury']
)
print("  - Created 'price_category': Quartile-based price categories")

# 7. Flag for ideal cut
df_clean['is_ideal_cut'] = (df_clean['cut'] == 'Ideal').astype(int)
print("  - Created 'is_ideal_cut': Binary flag for ideal cut diamonds")

# 8. Flag for best color (D, E, F are considered colorless/premium)
df_clean['is_premium_color'] = df_clean['color'].isin(['D', 'E', 'F']).astype(int)
print("  - Created 'is_premium_color': Binary flag for premium color grades (D, E, F)")

# 9. Flag for best clarity (IF, VVS1, VVS2 are considered flawless/near-flawless)
df_clean['is_premium_clarity'] = df_clean['clarity'].isin(['IF', 'VVS1', 'VVS2']).astype(int)
print("  - Created 'is_premium_clarity': Binary flag for premium clarity grades")

print(f"\nNew shape: {df_clean.shape}")
print(f"New columns: {df_clean.shape[1] - diamonds.shape[1]} additional columns created")


Creating derived columns...
  - Created 'price_per_carat': price divided by carat weight
  - Created 'volume': x * y * z (cubic mm)
  - Created 'depth_calculated': calculated depth percentage
  - 'table' column already exists as table percentage
  - Created 'size_category': Small (<0.5), Medium (0.5-1.0), Large (1.0-2.0), Very Large (>2.0)
  - Created 'price_category': Quartile-based price categories
  - Created 'is_ideal_cut': Binary flag for ideal cut diamonds
  - Created 'is_premium_color': Binary flag for premium color grades (D, E, F)
  - Created 'is_premium_clarity': Binary flag for premium clarity grades

New shape: (53940, 18)
New columns: 8 additional columns created


In [12]:
# Display the new columns
print("New columns created:")
new_cols = [col for col in df_clean.columns if col not in diamonds.columns]
print(new_cols)
print("\nSample of new columns:")
df_clean[['carat', 'price', 'price_per_carat', 'size_category', 'price_category', 
          'is_ideal_cut', 'is_premium_color', 'is_premium_clarity']].head(10)


New columns created:
['price_per_carat', 'volume', 'depth_calculated', 'size_category', 'price_category', 'is_ideal_cut', 'is_premium_color', 'is_premium_clarity']

Sample of new columns:


,carat,price,price_per_carat,size_category,price_category,is_ideal_cut,is_premium_color,is_premium_clarity
0,0.23,326,1417.391304,Small,Budget,1,1,0
1,0.21,326,1552.380952,Small,Budget,0,1,0
2,0.23,327,1421.739130,Small,Budget,0,1,0
3,0.29,334,1151.724138,Small,Budget,0,0,0
4,0.31,335,1080.645161,Small,Budget,0,0,0
5,0.24,336,1400.000000,Small,Budget,0,0,1
6,0.24,336,1400.000000,Small,Budget,0,0,1
7,0.26,337,1296.153846,Small,Budget,0,0,0
8,0.22,337,1531.818182,Small,Budget,0,1,0
9,0.23,338,1469.565217,Small,Budget,0,0,0


In [13]:
# Summary by cut quality
print("=" * 60)
print("SUMMARY BY CUT QUALITY")
print("=" * 60)
cut_summary = df_clean.groupby('cut').agg({
    'price': ['mean', 'median', 'count'],
    'carat': 'mean',
    'price_per_carat': 'mean'
}).round(2)
cut_summary.columns = ['Avg Price', 'Median Price', 'Count', 'Avg Carat', 'Avg Price/Carat']
cut_summary


SUMMARY BY CUT QUALITY


/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/3780052256.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  cut_summary = df_clean.groupby('cut').agg({


,Avg Price,Median Price,Count,Avg Carat,Avg Price/Carat
cut,,,,,
Fair,4358.76,3282.0,1610,1.05,3767.26
Good,3928.86,3050.5,4906,0.85,3860.03
Very Good,3981.76,2648.0,12082,0.81,4014.13
Premium,4584.26,3185.0,13791,0.89,4222.91
Ideal,3457.54,1810.0,21551,0.70,3919.70


In [14]:
# Summary by color grade
print("=" * 60)
print("SUMMARY BY COLOR GRADE")
print("=" * 60)
color_summary = df_clean.groupby('color').agg({
    'price': ['mean', 'median', 'count'],
    'carat': 'mean',
    'price_per_carat': 'mean'
}).round(2)
color_summary.columns = ['Avg Price', 'Median Price', 'Count', 'Avg Carat', 'Avg Price/Carat']
color_summary


SUMMARY BY COLOR GRADE


/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/564444632.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  color_summary = df_clean.groupby('color').agg({


,Avg Price,Median Price,Count,Avg Carat,Avg Price/Carat
color,,,,,
J,5323.82,4234.0,2808,1.16,3825.65
I,5091.87,3730.0,5422,1.03,3996.40
H,4486.67,3460.0,8304,0.91,4008.03
G,3999.14,2242.0,11292,0.77,4163.41
F,3724.89,2343.5,9542,0.74,4134.73
E,3076.75,1739.0,9797,0.66,3804.61
D,3169.95,1838.0,6775,0.66,3952.56


In [15]:
# Summary by clarity grade
print("=" * 60)
print("SUMMARY BY CLARITY GRADE")
print("=" * 60)
clarity_summary = df_clean.groupby('clarity').agg({
    'price': ['mean', 'median', 'count'],
    'carat': 'mean',
    'price_per_carat': 'mean'
}).round(2)
clarity_summary.columns = ['Avg Price', 'Median Price', 'Count', 'Avg Carat', 'Avg Price/Carat']
clarity_summary


SUMMARY BY CLARITY GRADE


/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/2352674228.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  clarity_summary = df_clean.groupby('clarity').agg({


,Avg Price,Median Price,Count,Avg Carat,Avg Price/Carat
clarity,,,,,
I1,3924.17,3344.0,741,1.28,2796.30
SI2,5063.03,4072.0,9194,1.08,4010.85
SI1,3996.00,2822.0,13065,0.85,3849.08
VS2,3924.99,2054.0,12258,0.76,4080.53
VS1,3839.46,2005.0,8171,0.73,4155.82
VVS2,3283.74,1311.0,5066,0.60,4204.17
VVS1,2523.11,1093.0,3655,0.50,3851.41
IF,2864.84,1080.0,1790,0.51,4259.93


In [16]:
# Summary by size category
print("=" * 60)
print("SUMMARY BY SIZE CATEGORY")
print("=" * 60)
size_summary = df_clean.groupby('size_category').agg({
    'price': ['mean', 'median', 'count'],
    'carat': ['mean', 'min', 'max'],
    'price_per_carat': 'mean'
}).round(2)
size_summary.columns = ['Avg Price', 'Median Price', 'Count', 'Avg Carat', 'Min Carat', 'Max Carat', 'Avg Price/Carat']
size_summary


SUMMARY BY SIZE CATEGORY


/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/210583656.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  size_summary = df_clean.groupby('size_category').agg({


,Avg Price,Median Price,Count,Avg Carat,Min Carat,Max Carat,Avg Price/Carat
size_category,,,,,,,
Small,839.72,788.0,18932,0.35,0.20,0.50,2336.45
Medium,2811.34,2528.0,17506,0.72,0.51,1.00,3770.18
Large,7607.72,6661.0,15613,1.26,1.01,2.00,5941.20
Very Large,14951.25,15392.0,1889,2.15,2.01,5.01,6997.50


In [17]:
# Pivot table: Average price by cut and color
print("=" * 60)
print("PIVOT TABLE: AVERAGE PRICE BY CUT AND COLOR")
print("=" * 60)
pivot_cut_color = pd.pivot_table(
    df_clean,
    values='price',
    index='cut',
    columns='color',
    aggfunc='mean'
).round(0)
pivot_cut_color


PIVOT TABLE: AVERAGE PRICE BY CUT AND COLOR


/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/2552842086.py:5: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_cut_color = pd.pivot_table(


color,J,I,H,G,F,E,D
cut,,,,,,,
Fair,4976.0,4685.0,5136.0,4239.0,3827.0,3682.0,4291.0
Good,4574.0,5079.0,4276.0,4123.0,3496.0,3424.0,3405.0
Very Good,5104.0,5256.0,4535.0,3873.0,3779.0,3215.0,3470.0
Premium,6295.0,5946.0,5217.0,4501.0,4325.0,3539.0,3631.0
Ideal,4918.0,4452.0,3889.0,3721.0,3375.0,2598.0,2629.0


In [18]:
# Pivot table: Average price per carat by cut and clarity
print("=" * 60)
print("PIVOT TABLE: AVERAGE PRICE PER CARAT BY CUT AND CLARITY")
print("=" * 60)
pivot_cut_clarity = pd.pivot_table(
    df_clean,
    values='price_per_carat',
    index='cut',
    columns='clarity',
    aggfunc='mean'
).round(0)
pivot_cut_clarity


PIVOT TABLE: AVERAGE PRICE PER CARAT BY CUT AND CLARITY


/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/2252804640.py:5: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_cut_clarity = pd.pivot_table(


clarity,I1,SI2,SI1,VS2,VS1,VVS2,VVS1,IF
cut,,,,,,,,
Fair,2409.0,3850.0,3881.0,4126.0,4097.0,4226.0,4804.0,3941.0
Good,2733.0,3790.0,3648.0,4195.0,4056.0,3976.0,3539.0,5073.0
Very Good,2949.0,4026.0,3824.0,4205.0,4104.0,3939.0,3695.0,5399.0
Premium,2811.0,4162.0,4044.0,4350.0,4444.0,4486.0,4008.0,4850.0
Ideal,3288.0,3948.0,3776.0,3814.0,4042.0,4260.0,3885.0,3851.0


In [19]:
# Value counts for categorical variables
print("=" * 60)
print("VALUE COUNTS FOR CATEGORICAL VARIABLES")
print("=" * 60)

print("\nCut distribution:")
print(df_clean['cut'].value_counts().sort_index())

print("\nColor distribution:")
print(df_clean['color'].value_counts().sort_index())

print("\nClarity distribution:")
print(df_clean['clarity'].value_counts().sort_index())

print("\nSize category distribution:")
print(df_clean['size_category'].value_counts().sort_index())

print("\nPrice category distribution:")
print(df_clean['price_category'].value_counts())


VALUE COUNTS FOR CATEGORICAL VARIABLES

Cut distribution:
cut
Fair          1610
Good          4906
Very Good    12082
Premium      13791
Ideal        21551
Name: count, dtype: int64

Color distribution:
color
J     2808
I     5422
H     8304
G    11292
F     9542
E     9797
D     6775
Name: count, dtype: int64

Clarity distribution:
clarity
I1        741
SI2      9194
SI1     13065
VS2     12258
VS1      8171
VVS2     5066
VVS1     3655
IF       1790
Name: count, dtype: int64

Size category distribution:
size_category
Small         18932
Medium        17506
Large         15613
Very Large     1889
Name: count, dtype: int64

Price category distribution:
price_category
Mid-Range    13495
Budget       13490
Luxury       13485
Premium      13470
Name: count, dtype: int64


In [20]:
# Cross-tabulation: Cut vs Color
print("=" * 60)
print("CROSS-TABULATION: CUT vs COLOR")
print("=" * 60)
pd.crosstab(df_clean['cut'], df_clean['color'], margins=True)


CROSS-TABULATION: CUT vs COLOR


color,J,I,H,G,F,E,D,All
cut,,,,,,,,
Fair,119,175,303,314,312,224,163,1610
Good,307,522,702,871,909,933,662,4906
Very Good,678,1204,1824,2299,2164,2400,1513,12082
Premium,808,1428,2360,2924,2331,2337,1603,13791
Ideal,896,2093,3115,4884,3826,3903,2834,21551
All,2808,5422,8304,11292,9542,9797,6775,53940


In [21]:
# Summary statistics for premium flags
print("=" * 60)
print("PREMIUM FEATURES SUMMARY")
print("=" * 60)
premium_summary = pd.DataFrame({
    'Ideal Cut': [df_clean['is_ideal_cut'].sum(), f"{df_clean['is_ideal_cut'].mean()*100:.1f}%"],
    'Premium Color (D/E/F)': [df_clean['is_premium_color'].sum(), f"{df_clean['is_premium_color'].mean()*100:.1f}%"],
    'Premium Clarity (IF/VVS)': [df_clean['is_premium_clarity'].sum(), f"{df_clean['is_premium_clarity'].mean()*100:.1f}%"]
}, index=['Count', 'Percentage'])
premium_summary


PREMIUM FEATURES SUMMARY


,Ideal Cut,Premium Color (D/E/F),Premium Clarity (IF/VVS)
Count,21551,26114,10511
Percentage,40.0%,48.4%,19.5%


## 6. Deeper Analysis: Price per Carat Controlling for Size

The previous summaries showed counterintuitive patterns because better color/clarity diamonds tend to be smaller. Let's analyze price per carat while controlling for size to see the true relationship.


In [22]:
# Price per carat by color, controlling for size category
print("=" * 60)
print("PRICE PER CARAT BY COLOR (CONTROLLING FOR SIZE)")
print("=" * 60)
color_size_price = df_clean.groupby(['color', 'size_category'])['price_per_carat'].mean().unstack()
print(color_size_price.round(2))
print("\nNote: Within each size category, better color (D/E/F) should show higher price per carat")


PRICE PER CARAT BY COLOR (CONTROLLING FOR SIZE)
size_category    Small   Medium    Large  Very Large
color                                               
J              1637.66  2764.03  4452.87     6310.44
I              1850.72  3200.93  5051.07     7075.29
H              2091.94  3528.84  5356.36     7086.05
G              2372.78  3754.21  6494.17     7159.12
F              2451.07  3902.93  6620.71     7491.27
E              2467.25  3941.10  6517.80     7670.30
D              2492.03  4169.37  6706.61     7623.91

Note: Within each size category, better color (D/E/F) should show higher price per carat


/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/1971612017.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  color_size_price = df_clean.groupby(['color', 'size_category'])['price_per_carat'].mean().unstack()


In [23]:
# Price per carat by clarity, controlling for size category  
print("=" * 60)
print("PRICE PER CARAT BY CLARITY (CONTROLLING FOR SIZE)")
print("=" * 60)
clarity_size_price = df_clean.groupby(['clarity', 'size_category'])['price_per_carat'].mean().unstack()
print(clarity_size_price.round(2))
print("\nNote: Within each size category, better clarity (IF/VVS) should show higher price per carat")


PRICE PER CARAT BY CLARITY (CONTROLLING FOR SIZE)
size_category    Small   Medium     Large  Very Large
clarity                                              
I1             1563.54  2087.72   3115.56     3533.69
SI2            1729.48  3279.40   4563.98     6890.08
SI1            1955.62  3548.30   5419.39     7455.01
VS2            2278.36  3807.88   6483.31     7608.34
VS1            2298.09  4034.87   7002.18     7711.82
VVS2           2571.07  4520.72   8361.02     8032.73
VVS1           2749.64  4698.63   9096.13     7808.19
IF             2970.96  5089.99  10028.84     8326.67

Note: Within each size category, better clarity (IF/VVS) should show higher price per carat


/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/3708451670.py:5: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  clarity_size_price = df_clean.groupby(['clarity', 'size_category'])['price_per_carat'].mean().unstack()


In [24]:
# Focus on medium-sized diamonds (0.5-1.0 carat) where most data is
print("=" * 60)
print("ANALYSIS FOR MEDIUM DIAMONDS (0.5-1.0 carat)")
print("=" * 60)
medium_diamonds = df_clean[(df_clean['carat'] >= 0.5) & (df_clean['carat'] <= 1.0)]
print(f"Number of diamonds in this range: {len(medium_diamonds):,} ({len(medium_diamonds)/len(df_clean)*100:.1f}% of total)")

print("\nPrice per carat by color (0.5-1.0 carat range):")
color_medium = medium_diamonds.groupby('color')['price_per_carat'].agg(['mean', 'count']).round(2)
color_medium.columns = ['Avg Price/Carat', 'Count']
print(color_medium.sort_index())


ANALYSIS FOR MEDIUM DIAMONDS (0.5-1.0 carat)
Number of diamonds in this range: 18,764 (34.8% of total)

Price per carat by color (0.5-1.0 carat range):
       Avg Price/Carat  Count
color                        
J              2737.27    720
I              3174.58   1470
H              3493.92   2359
G              3713.91   3664
F              3843.55   3771
E              3858.17   3991
D              4089.88   2789


/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/859212450.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  color_medium = medium_diamonds.groupby('color')['price_per_carat'].agg(['mean', 'count']).round(2)


In [25]:
print("\nPrice per carat by clarity (0.5-1.0 carat range):")
clarity_medium = medium_diamonds.groupby('clarity')['price_per_carat'].agg(['mean', 'count']).round(2)
clarity_medium.columns = ['Avg Price/Carat', 'Count']
print(clarity_medium.sort_index())



Price per carat by clarity (0.5-1.0 carat range):
         Avg Price/Carat  Count
clarity                        
I1               2043.90    222
SI2              3231.45   3270
SI1              3488.32   5305
VS2              3751.96   4489
VS1              3982.24   2831
VVS2             4459.95   1495
VVS1             4641.67    836
IF               5049.24    316


/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/3406835888.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  clarity_medium = medium_diamonds.groupby('clarity')['price_per_carat'].agg(['mean', 'count']).round(2)


In [26]:
# Visual comparison: Average carat size by color and clarity
print("=" * 60)
print("AVERAGE CARAT SIZE BY COLOR AND CLARITY")
print("=" * 60)
print("\nThis shows why the unadjusted averages were misleading:")
print("\nAverage carat by color:")
print(df_clean.groupby('color')['carat'].mean().sort_index().round(3))
print("\nAverage carat by clarity:")
print(df_clean.groupby('clarity')['carat'].mean().sort_index().round(3))


AVERAGE CARAT SIZE BY COLOR AND CLARITY

This shows why the unadjusted averages were misleading:

Average carat by color:


/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/3524461991.py:7: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df_clean.groupby('color')['carat'].mean().sort_index().round(3))
/var/folders/m8/6pxbwc7d15vbv2njbgcyjvxh0000gn/T/ipykernel_71985/3524461991.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df_clean.groupby('clarity')['carat'].mean().sort_index().round(3))


color
J    1.162
I    1.027
H    0.912
G    0.771
F    0.737
E    0.658
D    0.658
Name: carat, dtype: float64

Average carat by clarity:
clarity
I1      1.284
SI2     1.078
SI1     0.850
VS2     0.764
VS1     0.727
VVS2    0.596
VVS1    0.503
IF      0.505
Name: carat, dtype: float64


## Summary of Data Modification Steps

### Steps Completed:

1. **Data Loading**: Loaded the diamonds dataset using Seaborn (53,940 rows, 10 columns)

2. **Exploratory Data Analysis**:
   - Examined dataset shape, data types, and structure
   - Checked for missing values (none found)
   - Analyzed unique values for categorical columns (cut, color, clarity)
   - Reviewed numerical summary statistics

3. **Data Cleaning**:
   - Set proper data types: Converted categorical columns to ordered categories
     - Cut: Fair < Good < Very Good < Premium < Ideal
     - Color: J (worst) to D (best)
     - Clarity: I1 (worst) to IF (best)
   - Checked for outliers (found some, but kept them as they may be legitimate)
   - No missing values to handle

4. **Data Transformation**:
   - Created derived columns:
     - `price_per_carat`: Price divided by carat weight
     - `volume`: x * y * z (cubic mm)
     - `depth_calculated`: Calculated depth percentage
     - `size_category`: Binned carat into Small, Medium, Large, Very Large
     - `price_category`: Quartile-based price categories (Budget, Mid-Range, Premium, Luxury)
     - `is_ideal_cut`: Binary flag for ideal cut diamonds
     - `is_premium_color`: Binary flag for premium color grades (D, E, F)
     - `is_premium_clarity`: Binary flag for premium clarity grades (IF, VVS1, VVS2)

5. **Data Summarization**:
   - Groupby aggregations by cut, color, clarity, and size category
   - Pivot tables showing relationships between cut/color and cut/clarity
   - Value counts for all categorical variables
   - Cross-tabulations for cut vs color
   - Summary of premium features

The cleaned and transformed dataset is now ready for visualization and further analysis.
